Basic documents and DB CRUD commands

sashatim125 opened the pull request.

Commit:
update

Commit:
some minor changes to the commands.h (more like the java client)

Commit:
some changes to the commands.h(cpp) (more like the java client)

Commit:
update

Commit:
update - exceptions

Commit:
exceptions update - running

Commit:
PutDocumentCommand

Commit:
PutDocumentCommand
GetDatabaseRecordCommand
CreateDatabaseCommand
 can create db,retrieve it's record, put a document and retrieve it. :)

Commit:
DeleteDocumentCommand

Commit:
Google Tests Framework
First 2 tests on RequestExecutor

Commit:
GetDatabaseNamesCommand.h
Tryouts project
using wrapper types for from_json (instead of basic types,like long)

Commit:
DeleteDatabaseCommand
more tests

Commit:
minor changes

Commit:
DocumentCommandsTests

Commit:
refactoring : using namespaces,
each type in single file with relevant (de)serialization functions

ayende commented on Raven.CppClient.Tests/DocumentCommandsTests.cpp:
```diff
@@ -0,0 +1,182 @@
+
+namespace
+{
+	const std::string RAVEN_SERVER_URL = "http://localhost:8080";
```
Let's define all of these values in a single file / location.
That includes constants like the server url but also the entities / serialization code.
It shouldn't be in the same file as the tests

ayende commented on Raven.CppClient.Tests/DocumentCommandsTests.cpp:
```diff
@@ -0,0 +1,182 @@
+{
+	static bool contains_test_prefix(const std::string& str)
+	{
+		return (str.compare(0, test_prefix.length(), test_prefix) == 0);
```
what is this is too small to be a prefix?

ayende commented on Raven.CppClient.Tests/DocumentCommandsTests.cpp:
```diff
@@ -0,0 +1,182 @@
+	static void TearDownTestCase() //delete all TEST DBs
+	{
+		auto executor = http::RequestExecutor::create({ RAVEN_SERVER_URL }, {});
+		serverwide::operations::GetDatabaseNamesCommand cmd(0, 100);
```
Why are you deleting all the databases?
You should only delete the databases _for that particular test_

ayende replied:
in `SetupTestCase`, add new db name to a vector ` clear it in the `TeadDownTestCase`

ayende replied:
The reason for this is that we want to be able to run these tests concurrently at one point, and that would prevent that.

ayende commented on Raven.CppClient.Tests/DocumentCommandsTests.cpp:
```diff
@@ -0,0 +1,182 @@
+};
+
+const std::string DocumentCommandsTests::test_prefix{ "TEST__" };
+const std::string DocumentCommandsTests::db{ "TEST__DocumentCommandsTests" };
```
We don't need any special prefix

ayende replied:
we can use  "__func__ -__LINE" to generate unique db name more easily, no?

ayende commented on Raven.CppClient.Tests/DocumentCommandsTests.cpp:
```diff
@@ -0,0 +1,182 @@
+
+static bool does_document_exist(const std::string& db_name, const std::string& doc_id)
+{
+	auto executor = http::RequestExecutor::create({ RAVEN_SERVER_URL }, db_name);
```
Any reason you keep creating new request executors? The common use case (and certainly the one we want to test) it to hold to one and use it multiple times.

ayende commented on Raven.CppClient.Tests/DocumentCommandsTests.cpp:
```diff
@@ -0,0 +1,182 @@
+		{
+			continue;
+		}
+		if (doc.at("@metadata").at("@id").get<std::string>() == doc_id)
```
This will fail for case insensitivity

sashatim125 replied:
yes.
it seems like that comparison  was  unnecessary , since if the document exists,
it is returned, otherwise null is returned .
the order of documents (or nulls) is matched to the order of ids in the request

ayende commented on Raven.CppClient.Tests/DocumentCommandsTests.cpp:
```diff
@@ -0,0 +1,182 @@
+	bool docExists = false;
+	for (auto& doc : res.results)
+	{
+		if (std::string(doc.type_name()) == "null")
```
doc is a json token here?

ayende commented on Raven.CppClient.Tests/DocumentCommandsTests.cpp:
```diff
@@ -0,0 +1,182 @@
+	nlohmann::json j = user;
+	std::string id{ "newID" };
+	auto executor = http::RequestExecutor::create({ RAVEN_SERVER_URL }, db);
+	{
```
why the scope here?

ayende commented on Raven.CppClient.Tests/DocumentCommandsTests.cpp:
```diff
@@ -0,0 +1,182 @@
+	documents::commands::GetDocumentsCommand cmd(id, {}, false);
+	auto&& res = executor->execute(cmd);
+	bool doc_exists = false;
+	for (auto& doc : res.results)
```
This should be replaced with a call to `does_document_exist`

ayende commented on Raven.CppClient.Tests/RequestExecutorAndDBsTests.cpp:
```diff
@@ -0,0 +1,143 @@
+
+namespace
+{
+	const std::string RAVEN_SERVER_URL = "http://localhost:8080";
```
Same as before, we want to have _one_ such definition for the tests.

ayende commented on Raven.CppClient.Tests/RequestExecutorAndDBsTests.cpp:
```diff
@@ -0,0 +1,143 @@
+public:
+	static const std::string test_prefix;
+	static const std::string cgdr_db;
+	static const std::string cdd_db;
```
what is the meaning of `cdd` and `cgdr`?

ayende commented on Raven.CppClient.Tests/RequestExecutorAndDBsTests.cpp:
```diff
@@ -0,0 +1,143 @@
+
+class RequestExecutorTests : public ::testing::Test
+{
+	static bool contains_test_prefix(const std::string& str)
```
same as previous, and probably not required

ayende commented on Raven.CppClient.Tests/RequestExecutorAndDBsTests.cpp:
```diff
@@ -0,0 +1,143 @@
+		auto executor = http::RequestExecutor::create({ RAVEN_SERVER_URL }, {});
+		serverwide::operations::GetDatabaseNamesCommand cmd(0, 100);
+		auto&& res = executor->execute<std::vector<std::string>>(cmd);
+		for(const auto& dbname : res)
```
See previous test, you shouldn't be writing this but know, upfront, what dbs you delete

ayende commented on Raven.CppClient.Tests/RequestExecutorAndDBsTests.cpp:
```diff
@@ -0,0 +1,143 @@
+};
+
+const std::string RequestExecutorTests::test_prefix{ "TEST__" };
+const std::string RequestExecutorTests::cgdr_db{ "TEST__CanGetDatabaseRecord" };
```
Use `CreateDatabase(__func__)` instead to generate dbs

ayende commented on Raven.CppClient.Tests/RequestExecutorAndDBsTests.cpp:
```diff
@@ -0,0 +1,143 @@
+}
+
+TEST_F(RequestExecutorTests, ThrowsWhenDatabaseDoesNotExist)
+{//TODO refactor this for using exceptions and unsync 
```
unsync?

sashatim125 replied:
typo , means async 
anyway,corrected now

ayende commented on Raven.CppClient.Tests/RequestExecutorAndDBsTests.cpp:
```diff
@@ -0,0 +1,143 @@
+	}
+	catch (std::exception& exception)
+	{
+		auto ex = dynamic_cast<RavenError*>(&exception);
```
Can't you just catch `RavenError` directly?

sashatim125 replied:
yes. done like your said

ayende commented on Raven.CppClient.Tests/RequestExecutorAndDBsTests.cpp:
```diff
@@ -0,0 +1,143 @@
+	ASSERT_EQ(res.name, rec.database_name);
+}
+
+
```
Trim the file please

sashatim125 replied:
done

ayende commented on Raven.CppClient/ConflictSolver.h:
```diff
@@ -0,0 +1,12 @@
+	struct ConflictSolver
+	{
+		std::map<std::string, ScriptResolver> resolve_by_collection{};
+		bool resolve_to_latest = true;//am I right ?
```
`resolve_to_latest` is the default, yes.

sashatim125 replied:
OK

ayende commented on Raven.CppClient/CreateDatabaseCommand.h:
```diff
@@ -0,0 +1,63 @@
+	class CreateDatabaseCommand : public RavenCommand<DatabasePutResult>
+	{
+	private:
+		std::unique_ptr<DatabaseRecord> _database_record;//polymorphism is needed
```
why?

sashatim125 replied:
because (along with original c# code)  the base class DatabaseRecord is derived with DatabaseRecordWithEtag
(adding etag only)
```
    public class DatabaseRecordWithEtag : DatabaseRecord
    {
        public long Etag { get; set; }
    }
```
I think it's an overkill, and I propose to use DatabaseRecord only with (perhaps std::optional)
etag. What do you think ?

ayende replied:
Yes, go ahead

ayende commented on Raven.CppClient/CurlHandlesHolder.h:
```diff
@@ -0,0 +1,135 @@
+			CurlGlobalInit()
+			{
+				auto res = curl_global_init(CURL_GLOBAL_ALL);
+				if (res not_eq CURLE_OK)
```
why `not_eq` ?

sashatim125 replied:
checking if the initialization succeeded 
(now replaced with != )

ayende commented on Raven.CppClient/CurlHandlesHolder.h:
```diff
@@ -0,0 +1,135 @@
+			ReturnCurl& operator=(ReturnCurl&& other) = delete;
+
+			ReturnCurl(ReturnCurl&& other) = delete;/*noexcept
+				: _curl(std::move(other._curl))
```
remove commented code

sashatim125 replied:
done

ayende commented on Raven.CppClient/DatabasePutResult.h:
```diff
@@ -0,0 +1,23 @@
+	{
+		dpr.raft_command_index = j.at("RaftCommandIndex").get<int64_t>();
+		dpr.name = j.at("Name").get<std::string>();
+		//dpr.topology = j.at("Topology").get<DatabaseTopology>();
```
why is this commented?

sashatim125 replied:
because the DatabaseTopology deserialization (from_json) routine was not yes written.
it is now, but should be tested because of it's complexity

ayende commented on Raven.CppClient/DatabaseTopology.h:
```diff
@@ -0,0 +1,28 @@
+
+	struct DatabaseTopology
+	{
+		std::vector<std::string> members{};
```
can't we define this just as `;` without `{};`?

sashatim125 replied:
surely we can !
but it's a safety habit , a part of the wider one "do not leave members uninitialized".
also explicitly calling default ctor when it doesn't exist 
results in compilation error , that way reminding to make
the initialization in the non_default ctor.

ayende commented on Raven.CppClient/DeleteDatabaseCommand.h:
```diff
@@ -0,0 +1,97 @@
+
+namespace ravendb::client::serverwide::operations
+{
+	using TimeToWaitForConfirmation_t = std::chrono::milliseconds; //which units should we use ?
```
We use milliseconds
But I don't like the `TimeToWait..` it actually obscures what is going on. Use `std::chrono::milliseconds` directly

sashatim125 replied:
done

ayende commented on Raven.CppClient/DeleteDatabaseCommand.h:
```diff
@@ -0,0 +1,97 @@
+	class DeleteDatabaseCommand : public RavenCommand<DeleteDatabaseResult>
+	{
+	private:
+		DeleteDatabaseParameters _parameters{};
```
why parameters and not have this directly on this class?

sashatim125 replied:
it's the java code design, 
where this command is actually wrapped by DeleteDatabasesOperation,
so the Parameters object is a block that is passed by the operation 
to the command upon it's creation (by getCommand() )
so in my code this ctor
```
DeleteDatabaseCommand(std::string database_name, bool hard_delete, std::string from_node = {},
			std::chrono::milliseconds time_to_wait_for_confirmation = default_wait_time)
```
will be move to operation and only this one
`DeleteDatabaseCommand(DeleteDatabaseParameters parameters)`
will be left.

ayende commented on Raven.CppClient/DeleteDocumentCommand.h:
```diff
@@ -0,0 +1,62 @@
+				std::ostringstream change_vector_header;
+				change_vector_header << "If-Match:" << '"' << _change_vector << '"';
+
+				_headers_list = curl_slist_append(_headers_list, change_vector_header.str().c_str());
```
who owns the memory in `_headers_list` ?

sashatim125 replied:
curl_slist* is a pointer to linked list's head, while the list is allocated on the heap
so  curl_slist_append returns a new head (or NULL)
(so my code wasn't safe line 51).
curl_easy_setopt(curl, CURLOPT_HTTPHEADER, _headers_list) makes
the curl_easy_perform() use this list, i.e. it's used beyond the function where it
was created , so RAII is unappropriate

ayende replied:
So the issue is value you appened (the `change_vector_header`) is copied or referenced?

sashatim125 replied:
copied , as mentioned [here](https://curl.haxx.se/libcurl/c/curl_slist_append.html) :

> curl_slist_append appends a string to a linked list of strings. The existing list should be passed as the first argument and the new list is returned from this function. Pass in NULL in the list argument to create a new list. The specified string has been appended when this function returns. curl_slist_append **copies** the string.
> 
> The list should be freed again (after usage) with curl_slist_free_all.

ayende commented on Raven.CppClient/GetDatabaseNamesCommand.h:
```diff
@@ -0,0 +1,72 @@
+{
+	namespace
+	{
+		struct DatabaseNames_t//used with to/from json
```
We need to make a consistent decision on class names.
`DatabaseNames_t` is ugly in my eyes.
I think that class names should be `snake_case` to match the common behavior from the library.

@maximburyak thoughts?

maximburyak replied:
We discussed it with @sashatim125 and it seems best to go the camelcase route. This way code will be easier to port, readable to non cpp guys and more explorable and comparable to other clients, which all written using variations of camelCase

ayende replied:
camelCase route isn't common in C++
I would rather go with the std lib naming conventions

ayende commented on Raven.CppClient/GetDatabaseNamesCommand.h:
```diff
@@ -0,0 +1,72 @@
+		{
+			auto&& res = result.value;
+			auto&& it = j.find("Databases");
+			if (it == j.end() or not it->is_array())
```
`or not` ? 
Please don't use the alternative operators.

Also, isn't an empty array valid response here?

sashatim125 replied:
> Also, isn't an empty array valid response here?

it sure is ,like this 

> {
"Databases": [ ]
}

but absence of the "Databases" key in return json
or if it's value isn't an array is an error

ayende commented on Raven.CppClient/GetDatabaseNamesCommand.h:
```diff
@@ -0,0 +1,72 @@
+		{
+			long statusCode;
+			curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &statusCode);
+			if (statusCode == 200)
```
What about error handling when the result is not 200?

ayende commented on Raven.CppClient/GetDatabaseRecordCommand.h:
```diff
@@ -0,0 +1,46 @@
+		{
+			long statusCode;
+			curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &statusCode);
+			if (statusCode == 200)
```
What about error handling when the result is not 200?

ayende commented on Raven.CppClient/GetDatabaseTopologyCommand.h:
```diff
@@ -0,0 +1,46 @@
+			{
+				_result = response;
+			}
+			else
```
What about error handling when the result is not 200 / 404? What happens if you don't have authorization, for example.

ayende commented on Raven.CppClient/GetDocumentsCommand.h:
```diff
@@ -0,0 +1,179 @@
+				}
+			}
+			else // ids too big, must use POST
+			{
```
Note you need to handle this as well

https://github.com/ravendb/ravendb/commit/f0366efc7361b7c45cc0b3fcffe4b1d932f64f41

sashatim125 replied:
done 
(unused now, though, because we don't have the http cash yet)

ayende commented on Raven.CppClient/GetDocumentsCommand.h:
```diff
@@ -0,0 +1,179 @@
+		{
+			long statusCode;
+			curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &statusCode);
+			if (statusCode == 200)
```
error handling

ayende commented on Raven.CppClient/RavenTest.cpp:
```diff
@@ -0,0 +1,6 @@
+#include "gtest/gtest.h"
+
+TEST(TestCaseName, TestName) {
```
can remove this

sashatim125 replied:
done

ayende commented on Raven.CppClient/RequestExecutor.cpp:
```diff
@@ -0,0 +1,117 @@
+
+		rq->validate_urls();
+
+		if (not rq->_db_name.empty())
```
This is actually a lot more readable than `!`, maybe just allow `not`, @maximburyak?

maximburyak replied:
I prefere not using alternative logical operators at all. Bear in mind this should be easy to read to other team members as well

ayende commented on Raven.CppClient/RequestExecutor.h:
```diff
@@ -0,0 +1,145 @@
+				case 201:
+				{
+					auto result = nlohmann::json::parse(output_buffer);
+					cmd.set_response(curl, result, false);
```
Note that `set_response` is called _only_ for 200 / 2001, never for anything else.
That means that any code in `set_response` that try to handle that shouldn't.

sashatim125 replied:
OK, unnecessary code removed

sashatim125 replied:
http response code for now is processed in RequestExecutor only

ayende commented on Raven.CppClient/RequestExecutor.h:
```diff
@@ -0,0 +1,145 @@
+		{
+			std::ostringstream errors;
+
+			std::shared_ptr<Topology> topology_local = std::atomic_load(&_topology);
```
Is this safe? Is this going to safely inc the ref count in multi threaded fashion?

sashatim125 replied:
Yes.
according to [this](https://en.cppreference.com/w/cpp/memory/shared_ptr/atomic) :

> If multiple threads of execution access the **same** std::shared_ptr **object** without synchronization and any of those accesses uses a non-const member function of shared_ptr then a data race will occur unless all such access is performed through these functions, which are overloads of the corresponding atomic access functions (std::atomic_load, std::atomic_store, etc.)
> 
> Note that the control block of a shared_ptr is thread-safe: **different** std::shared_ptr **objects** can be accessed using mutable operations, such as operator= or reset, simultaneously by multiple threads, even when these instances are copies, and share the same control block internally.

ayende commented on Raven.CppClient/RequestExecutor.h:
```diff
@@ -0,0 +1,145 @@
+				}
+				catch (RavenError& re)
+				{
+					errors << re.what() << '\n';
```
Errors are rare, what is the cost of creating the `errors` on each call?

sashatim125 replied:
according to [this](https://en.cppreference.com/w/cpp/io/basic_stringbuf) not much,
but using std::optional seems like reasonable optimization

sashatim125 replied:
we can also keep the list of all exceptions(std::exception_ptr)
(probably as std::optional too)
and creates the errors string only when throwing

ayende replied:
If the cost is cheap, let's just leave it.

Commit:
updates

sashatim125 committed:
tests were refactored :
RequestExecutorAndDBsTests using RequestExecutorScope(r.e. + db),
DocumentCommandsTests with sample document creation per test.
errors correcting
